File: src/modifiers/dynamicTypeSize.ts
import { createModifier } from './createModifier';

/// SwiftUI `DynamicTypeSize` values, from `.xSmall` up through the five
/// accessibility sizes.
export type DynamicTypeSize =
  | 'xSmall'
  | 'small'
  | 'medium'
  | 'large'
  | 'xLarge'
  | 'xxLarge'
  | 'xxxLarge'
  | 'accessibility1'
  | 'accessibility2'
  | 'accessibility3'
  | 'accessibility4'
  | 'accessibility5';

/// SwiftUI `.dynamicTypeSize(_:)`. Pins this view's subtree to a fixed
/// Dynamic Type size rather than following the system text-size setting.
/// Requires watchOS 9+ — a no-op on older systems.
export function dynamicTypeSize(size: DynamicTypeSize) {
  return createModifier('dynamicTypeSize', { size });
}

File: apple/Sources/ReactNativeWatchOS/Modifiers/DynamicTypeSize.swift
import SwiftUI

/// SwiftUI `.dynamicTypeSize(_:)`. Pins the subtree to a fixed
/// `DynamicTypeSize` instead of tracking the system text-size setting.
/// Requires watchOS 9.0+ — returns the view unchanged on older systems.
enum RNWDynamicTypeSizeModifier {
    @MainActor
    static func register(into r: RNWModifierRegistry) {
        r.register("dynamicTypeSize") { view, params, _ in
            if #available(watchOS 9.0, *) {
                guard let size = parse(params.string("size")) else { return view }
                return AnyView(view.dynamicTypeSize(size))
            }
            return view
        }
    }

    @available(watchOS 9.0, *)
    private static func parse(_ s: String?) -> DynamicTypeSize? {
        switch s {
        case "xSmall":         return .xSmall
        case "small":          return .small
        case "medium":         return .medium
        case "large":          return .large
        case "xLarge":         return .xLarge
        case "xxLarge":        return .xxLarge
        case "xxxLarge":       return .xxxLarge
        case "accessibility1": return .accessibility1
        case "accessibility2": return .accessibility2
        case "accessibility3": return .accessibility3
        case "accessibility4": return .accessibility4
        case "accessibility5": return .accessibility5
        default:               return nil
        }
    }
}
